File: code/get_training_samples.py
"""
Creates positive training samples for distant supervision.

Usage:
    ./get_training_samples.py \
        --article_sentences_entities=allan_dwan.json \
        --article_sentences_entities=autism.json \
        ... \
        --output_file=x.json
"""

from google.protobuf import text_format
import relations
import file_util
import sentence_pb2
import training_samples_pb2
import wikidata

class MentionInSentence(object):
    def __init__(self, start_token_id, end_token_id, wikidata_id):
        self.start_token_id = start_token_id
        self.end_token_id = end_token_id
        self.wikidata_id = wikidata_id

class SentenceInDocument(object):
    def __init__(self, document, sentence_id):
        self.mentions = []
        wikidata_ids = set()
        for coreference in document.coreferences:
            if not coreference.wikidataEntityId:
                # entity not detected
                continue
            for mention in coreference.mentions:
                if mention.sentenceId == sentence_id:
                    self.mentions.append(MentionInSentence(mention.startWordId,
                                                           mention.endWordId,
                                                           coreference.wikidataEntityId))
                    wikidata_ids.add(coreference.wikidataEntityId)
        self.wikidata_ids = list(wikidata_ids)
        self.document = document
        self.sentence_id = sentence_id

        # print(self.get_text(), self.wikidata_ids)

    def all_entity_pairs(self):
        pairs = []
        for e1 in self.wikidata_ids:
            if e1 is None:
                # TODO: HAX SHOULD NOT BE NEEDED
                continue
            for e2 in self.wikidata_ids:
                if e2 is None:
                    # TODO: HAX SHOULD NOT BE NEEDED
                    continue
                pairs.append((e1, e2))
        return pairs

    def get_sentence(self):
        for sentence in self.document.sentences:
            if sentence.id == self.sentence_id:
                return sentence

    def get_text(self):
        return self.get_sentence().text

    def to_sample(self, relation, e1, e2, positive):
        assert (e1 in self.wikidata_ids) and (e2 in self.wikidata_ids)

        sample = training_samples_pb2.TrainingSample()
        sample.positive = positive
        sample.relation = relation
        sample.e1 = e1
        sample.e2 = e2

        sentence = self.get_sentence()

        for mention in self.mentions:
            for token_index in range(mention.start_token_id - 1, mention.end_token_id - 1):
                if mention.wikidata_id == e1:
                    sample.e1_token_indices.append(token_index)
                if mention.wikidata_id == e2:
                    sample.e2_token_indices.append(token_index)

        s = sample.sentence
        s.text = sentence.text
        sentence_start = sentence.tokens[0].start_offset
        for token in sentence.tokens:
            out_token = s.tokens.add()
            out_token.start_offset = token.start_offset - sentence_start
            out_token.end_offset = token.end_offset - sentence_start
            out_token.lemma = token.lemma
            out_token.pos = token.pos
            out_token.ner = token.ner

        return sample

def get_true_triples_expressed_by_sentence(sentence):
    """
    Returns:
        dict (key: "Rxxx", value: ("Q123", "Q456"))
    """
    mentioned_wikidata_ids = sentence.wikidata_ids
    sentence_entity_pairs = sentence.all_entity_pairs()
    wikidata_client = wikidata.WikidataClient()

    all_pairs = {}
    # print(mentioned_wikidata_ids)
    # print(sentence_entity_pairs)
    for wikidata_id in mentioned_wikidata_ids:
        for e1, rel, e2 in wikidata_client.get_all_triples_of_entity(wikidata_id):
            # print('(', e1, rel, e2, ')')
            key = (e1, e2)
            if key not in sentence_entity_pairs:
                # The relation holds, but the entity pair is in no sentences.
                continue
            if key not in all_pairs:
                # TODO: FIXME: relations are returned twice -- forward and backward
                all_pairs[key] = set()
            all_pairs[key].add(rel)
    return all_pairs

def sentence_to_training_data(sentence):
    """
    Args:
      sentence (SentenceInDocument)
    """
    print('sentence_to_training_data(', sentence.get_text(), ')')
    all_pairs = get_true_triples_expressed_by_sentence(sentence)

    print(sentence) #, all_pairs)
    #if len(all_pairs) > 0:
        #raise

    # Add positive samples for represented relations.
    # TODO: skip sentence if there are multiple candidate relations
    training_data = TrainingData()
    for entity_pair, true_relations in all_pairs.items():
        e1, e2 = entity_pair
        for relation in all_pairs[entity_pair]:
            print('\tpositive:', (e1, relation, e2))
            sample = sentence.to_sample(relation, e1, e2, positive=True)
            training_data.add_sample(sample)

    # TODO: add negative samples
    # Add negative samples for unrepresented relations.
    if len(sentence.wikidata_ids) >= 2:
        e1, e2 = sentence.wikidata_ids[:2]
        key = (e1, e2)
        for relation in relations.IMPORTANT_RELATIONS:
            if (key not in all_pairs) or (relation not in all_pairs[key]):
                print('\tnegative:', (e1, relation, e2))
                sample = sentence.to_sample(relation, e1, e2, positive=False)
                training_data.add_sample(sample)

    return training_data

def load_document(document):
    print('Loading document:', document.article_sanename, '...')
    sentences = []
    for sentence in document.sentences:
        # TODO: create more complex samples
        sentences.append(SentenceInDocument(document, sentence.id))
    return sentences

def load_document_files(files):
    """
    Returns:
      list of SentenceInDocument
    """
    sentences = []
    for path in files:
        document = file_util.parse_proto_file(sentence_pb2.Document, path)
        sentences.extend(load_document(document))
    return sentences

def join_sentences_entities(sentences):
    """
    Args:
      sentences (list of SentenceInDocument)
    """
    training_data = TrainingData()
    for sentence in sentences:
        sentence_training_data = sentence_to_training_data(sentence)
        training_data.add_training_data(sentence_training_data)
    return training_data

class TrainingData(object):
    def __init__(self):
        # key: relation id, value: list of samples, both positive and negative
        self.training_data = {}

    def load(self, path):
        td = file_util.parse_proto_file(
            training_samples_pb2.TrainingSamples,
            path)
        # TODO
        self.training_data = {}
        for relation_training_samples in td.relation_samples:
            self.training_data[relation_training_samples.relation] = (
                [sample for sample in relation_training_samples.samples])

    def add_sample(self, sample):
        if sample.relation not in self.training_data:
            self.training_data[sample.relation] = []
        # TODO: training data should also say where is the
        # relevant mention
        self.training_data[sample.relation].append(sample)

    def add_training_data(self, other):
        for relation, samples in other.training_data.items():
            if relation not in self.training_data:
                self.training_data[relation] = []
            self.training_data[relation].extend(samples)

    def to_proto(self):
        samples = training_samples_pb2.TrainingSamples()
        for relation, rs in self.training_data.items():
            rels = samples.relation_samples.add()
            rels.relation = relation
            rels.samples.extend(rs)
        return samples

    def write(self, output_file):
        samples = self.to_proto()
        # print(text_format.MessageToString(samples))
        with open(output_file, 'wb') as f:
            f.write(samples.SerializeToString())

File: code/hadoop/GetTrainingSamples.java
import java.util.List;
import java.util.ArrayList;
import java.util.Set;
import java.util.HashSet;
import java.util.Map;
import java.util.HashMap;
import org.apache.log4j.Logger;

public class GetTrainingSamples {
	static Logger log = Logger.getLogger(GetTrainingSamples.class);

	public static class MentionInSentence {
		public int startTokenId;
		public int endTokenId;
		public String wikidataId;

		public MentionInSentence(int startTokenId, int endTokenId, String wikidataId) {
			this.startTokenId = startTokenId;
			this.endTokenId = endTokenId;
			this.wikidataId = wikidataId;
		}
	}

	private static class EntityPair {
		public String e1;
		public String e2;

		public EntityPair(String e1, String e2) {
			this.e1 = e1;
			this.e2 = e2;
		}

		@Override
		public boolean equals(Object other) {
			if (!(other instanceof EntityPair)) {
				return false;
			}
			EntityPair o = (EntityPair) other;
			return e1.equals(o.e1) && e2.equals(o.e2);
		}

		@Override
		public int hashCode() {
			return e1.hashCode() ^ e2.hashCode();
		}

		@Override
		public String toString() {
			return "EntityPair{" + e1 + " " + e2 + "}";
		}
	}

	public static class SentenceInDocument {
		private List<MentionInSentence> mentions = new ArrayList<>();
		private List<String> wikidataIds = new ArrayList<>();
		private Sentence.Document document;
		private int sentenceId;

		public List<String> getWikidataIds() {
			return wikidataIds;
		}

		public SentenceInDocument(Sentence.Document document, int sentenceId) {
			this.document = document;
			this.sentenceId = sentenceId;

			Set<String> wikidataIds = new HashSet<>();
			for (Sentence.Coreference coref : document.getCoreferencesList()) {
				if (!coref.hasWikidataEntityId() || coref.getWikidataEntityId() == "") {
					// entity not detected
					continue;
				}
				for (Sentence.Mention mention : coref.getMentionsList()) {
					mentions.add(new MentionInSentence(
								mention.getStartWordId(),
								mention.getEndWordId(),
								coref.getWikidataEntityId()));
					wikidataIds.add(coref.getWikidataEntityId());
				}
			}
			this.wikidataIds = new ArrayList(wikidataIds);
		}

		public List<EntityPair> getAllEntityPairs() {
			List<EntityPair> pairs = new ArrayList<>();
			for (String e1 : wikidataIds) {
				if (e1 == null) {
					continue;
				}
				for (String e2 : wikidataIds) {
					if (e2 == null) {
						// TODO: hax should not be
						// needed
						continue;
					}
					pairs.add(new EntityPair(e1, e2));
				}
			}
			return pairs;
		}

		public Sentence.DocumentSentence getSentence() {
			for (Sentence.DocumentSentence sentence : document.getSentencesList()) {
				if (sentence.getId() == sentenceId) {
					return sentence;
				}
			}
			return null;
		}

		public String getText() {
			return getSentence().getText();
		}

		public TrainingSamples.TrainingSample toSample(String relation, String e1, String e2, boolean positive) {
			assert wikidataIds.contains(e1) && wikidataIds.contains(e2);

			TrainingSamples.TrainingSample.Builder sample = TrainingSamples.TrainingSample.newBuilder()
				.setPositive(positive)
				.setRelation(relation)
				.setE1(e1)
				.setE2(e2);

			Sentence.DocumentSentence sentence = getSentence();

			List<Integer> e1Indices = new ArrayList<>();
			List<Integer> e2Indices = new ArrayList<>();
			for (MentionInSentence mention : mentions) {
				for (int tokenIndex = mention.startTokenId - 1;
						tokenIndex < mention.endTokenId - 1;
						tokenIndex++) {
					if (mention.wikidataId.equals(e1) && !e1Indices.contains(tokenIndex)) {
						e1Indices.add(tokenIndex);
					}
					if (mention.wikidataId.equals(e2) && !e2Indices.contains(tokenIndex)) {
						e2Indices.add(tokenIndex);
					}
				}
			}
			sample.addAllE1TokenIndices(e1Indices)
				.addAllE2TokenIndices(e2Indices);

			TrainingSamples.TrainingSampleParsedSentence.Builder bldr = TrainingSamples.TrainingSampleParsedSentence.newBuilder();
			//sample.getSentence().setText(sentence.getText());
			bldr.setText(sentence.getText());

			int sentenceStart = sentence.getTokens(0).getStartOffset();
			for (Sentence.SentenceToken token : sentence.getTokensList()) {
				TrainingSamples.TrainingSampleSentenceToken.Builder b = TrainingSamples.TrainingSampleSentenceToken.newBuilder();
				b.setStartOffset(token.getStartOffset() - sentenceStart);
				b.setEndOffset(token.getEndOffset() - sentenceStart);
				b.setLemma(token.getLemma());
				b.setPos(token.getPos());
				b.setNer(token.getNer());
				bldr.addTokens(b);
			}
			sample.setSentence(bldr);
			return sample.build();
		}
	}

	private WikidataClient wikidataClient;

	public GetTrainingSamples(WikidataClient wikidataClient) {
		this.wikidataClient = wikidataClient;
	}

	public Map<EntityPair, List<String>> getTrueTriplesExpressedBySentence(SentenceInDocument sentence) {
		List<String> mentionedWikidataIds = sentence.getWikidataIds();
		List<EntityPair> sentenceEntityPairs = sentence.getAllEntityPairs();

		Map<EntityPair, List<String>> allPairs = new HashMap<>();
		for (String wikidataId : mentionedWikidataIds) {
			for (WikidataClient.Triple triple : wikidataClient.getAllTriplesOfEntity(wikidataId)) {
				EntityPair pair = new EntityPair(triple.subject, triple.object);
				if (!sentenceEntityPairs.contains(pair)) {
					// relation holds, but not expressed by sentence
					continue;
				}
				if (!allPairs.containsKey(pair)) {
					// TODO: fixme: relations returned twice -- forward and backward
					allPairs.put(pair, new ArrayList<>());
				}
				allPairs.get(pair).add(triple.predicate);
			}
		}
		return allPairs;
	}

	public List<TrainingSamples.TrainingSample> documentToSamples(Sentence.Document document) {
		ArrayList<TrainingSamples.TrainingSample> samples = new ArrayList<>();
		for (Sentence.DocumentSentence sentence : document.getSentencesList()) {
			SentenceInDocument sid = new SentenceInDocument(document, sentence.getId());
			samples.addAll(sentenceToTrainingSamples(sid));
		}
		return samples;
	}

	public List<TrainingSamples.TrainingSample> sentenceToTrainingSamples(SentenceInDocument sentence) {
		log.info("Sentence to training samples: " + sentence.getText());
		Map<EntityPair, List<String>> allPairs = getTrueTriplesExpressedBySentence(sentence);

		List<TrainingSamples.TrainingSample> samples = new ArrayList<>();
		// Add positive samples.
		for (EntityPair pair : allPairs.keySet()) {
			for (String relation : allPairs.get(pair)) {
				TrainingSamples.TrainingSample sample = sentence.toSample(relation, pair.e1, pair.e2, true);
				samples.add(sample);
			}
		}

		// TODO: add exactly as many positive samples as negative ones

		/*
		// Add negative samples.
		// TODO
		if (sentence.getWikidataIds().size() >= 2) {
			String e1 = sentence.getWikidataIds().get(0);
			String e2 = sentence.getWikidataIds().get(1);

			EntityPair key = new EntityPair(e1, e2);

			for (String relation : Relations.IMPORTANT_RELATIONS) {
				if ((!allPairs.containsKey(key)) || (!allPairs.get(key).contains(relation))) {
					TrainingSamples.TrainingSample sample = sentence.toSample(relation, e1, e2, false);
					samples.add(sample);
				}
			}
		}
		*/

		return samples;
	}
}
